README.md
# Information & Warning Popup for React.js.

## Installation

The recommended way to include this sdk into your project is by using npm. Install it into your project as a dependency with

```sh
$ npm install npup --save
```

## How to use

To include the modules in the browser context use a bundler tool like `Browserify` or `Webpack`.

## Import into your project

To include this project you need to require the module by using CommonJS syntax or ES6 Modules Syntax (recommended).


```js
// ES6
import NPupBars, { addMessage } from 'npup'
/* NPupBars is the default export.
   It represents the React Component that contains the NPupComponents.

   You can give the default export any name
   but in this example it will be referenced as NPupBars */

```

For best results, render the `NPupBars` component in the higher DOM node possible, such as:

```js
// react's app render function
render() {
	return (
		<div className='main'>
			<NPupBars />
			...
		</div>
	)
}

```

Then just call NPupBars with `addMessage(message)`

```js

addMessage('Info');

```

You can call NPupBars with `clearMessages()` to clear all messages.

```js

clearMessages();

```

## Options

The npup function `addMessage()` supports `text`, `closeText`, `closeCallback`, `type`, `timeout` and `color` attributes in the following way.

`addMessage(text, closeText, closeCallback, type, timeout, color)`


`text`: required => is the content of the npup.


`closeText`: not required => is the content of the close button caption of npup.


`closeCallback`: not required => is the callback function when close button clicked from outside.
if closeText did not define, it does not return callback.


`type`: not required => consists of three variants:

- `success` to render a success popup.
- `warning` to render a warning popup.
- `error` to render an error popup.
- `custom` to render user defined colors for the popup.

`timeout` is the time (in milliseconds) the toast will remain on screen.
if it's not set, it will display for the default `5000ms` time.
You can also pass `-1` to cause the notification to display persistently.

`color` is for the `background` as well as the `text` of the notification. It accepts an object with the following properties

```js
let myColorStyle = { background: '#0E1717', text: "#FFFFFF" };
addMessage("sample information", "", null, "custom", 4000, myColorStyle);
```

## Settings

Settings of npup

### Defaults

```js
import {NPupDefaults} from 'npup/defaults';
```

call `NPupDefaults.setDurations(defaultTimeout, firstAnimationDuration, lastAnimationDuration)` to set npup times.

`defaultTimeout`: required => is the time of show on the screen.

`firstAnimationDuration`: required => is the time of starting animation.

`lastAnimationDuration`: required => is the time of closing animation.

### Styles

```js
import {NPupStyles} from 'npup/styles';
```

call `NPupStyles.changeBackgroundColorStyle(newStyle)` to set background colors.

`newStyle`: required => is the background colors array.
```js
NPupStyles.changeBackgroundColorStyle({
	default: '#000000',
	warning: '#000000',
	error: '#000000',
	success: '#000000'
});
```

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Nusret Parlak](https://github.com/nusretparlak/).

## [Unreleased]
### Added

### Changed

- function of clearMessages() is added.

### Fixed

- webpack problem is fixed. 

src/index.js
import React from 'react';
import {NPupBar} from './npupBar';
import {NPupStyles} from './styles';
import {NPupDefaults} from './defaults';
import {NPupAnimations} from './animations';

/* Export addMessage function */
export function addMessage(text, closeText = '', closeCallback = null, type = '', timeout = NPupDefaults.defaultTimeout, color = NPupStyles.textColor.default) {
	let newMessageId = NPupDefaults.getNewElementId();
	NPupDefaults._NPupBar.refNPupBar.addMessage({messageId: newMessageId, text, closeText, closeCallback, type, timeout, color});
}

/* Export clearMessages function */
export function clearMessages() {
	NPupDefaults._NPupBar.refNPupBar.clearMessages();
}

/* Export container */
export default class NPupBars extends React.Component {
	shouldComponentUpdate() {
		return false;
	}

	render() {
		if (!NPupDefaults._NPupBar) {
			NPupDefaults.setNPupBar(this);
		}
		NPupAnimations.setAnimation();
		return (
			<NPupBar key={NPupDefaults.npupUniqueId} id={NPupDefaults.npupUniqueId} ref={(r) => this.refNPupBar = r} />
		);
	}
}

/* Export ChangeStyles function */
export function changeContainerStyle(newStyle) {
	NPupStyles.changeContainerStyle(newStyle);
}

export function changeContentContainerStyle(newStyle) {
	NPupStyles.changeContentContainerStyle(newStyle);
}

export function reRender() {
	NPupDefaults.reRenderNPupBar();
}


src/npupBar.js
import React from 'react';
import {NPupComponent} from './npupComponent';
import {NPupStyles} from './styles';
import {NPupDefaults} from './defaults';

/* React NPupComponent Component */
export class NPupBar extends React.Component {
	static queueLock = false;
	state = {
		messages: []
	};

	addMessage(newMessage) {
		NPupDefaults.addMessage(newMessage);
		this.setState({messages: Object.assign({}, NPupDefaults.messages)});
	}

	clearMessages() {
		NPupDefaults.clearMessages();
		this.setState({messages: Object.assign({}, NPupDefaults.messages)});
	}

	render() {
		let componentArray;
		if (NPupDefaults.messages) {
			componentArray = NPupDefaults.messages.map((message) => {
				return (
					<NPupComponent
						key={message.messageId}
						messageId={message.messageId}
						text={message.text}
						closeText={message.closeText}
						closeCallback={message.closeCallback}
						timeout={message.timeout}
						type={message.type}
						color={message.color} />
				);
			});
		}
		return (
			<div ref={(r) => this.mainDiv = r} className="npup" style={NPupStyles.containerStyle}>
				{componentArray}
			</div>
		);
	}
}


src/defaults.js
export class NPupDefaults {
	static npupUniqueId = 'npup-unique-id';
	static npupElementId = 'npup-element-';
	static firstAnimationName = 'npup-first-animation-name';
	static lastAnimationName = 'npup-last-animation-name';
	static _NPupBar;
	static messages = [];
	static firstAnimationDuration = 300; // ms
	static lastAnimationDuration = 500; // ms
	static defaultTimeout = 5000; // ms
	static elementId = 0;
	static closeCallback = null;

	static setDurations(defaultTimeout, firstAnimationDuration, lastAnimationDuration) {
		this.defaultTimeout = defaultTimeout;
		this.firstAnimationDuration = firstAnimationDuration;
		this.lastAnimationDuration = lastAnimationDuration;
	}
	static getNewElementId() {
		this.elementId += 1;
		return this.npupElementId + this.elementId;
	}
	static setNPupBar(__NPupBar) {
		this._NPupBar = __NPupBar;
	}
	static addMessage(newMessage) {
		this.messages.push(newMessage);
	}
	static removeMessage(messageId) {
		let deletedIndex = -1;
		for (let i = 0; i < this.messages.length; ++i) {
			if (this.messages[i].messageId === messageId) {
				deletedIndex = i;
				break;
			}
		}
		if (deletedIndex >= 0) {
			this.messages.splice(deletedIndex, 1);
			return true;
		}
		return false;
	}
	static clearMessages() {
		this.messages = [];
	}
	static isMessageExist(messageId) {
		let deletedIndex = -1;
		for (let i = 0; i < this.messages.length; ++i) {
			if (this.messages[i].messageId === messageId) {
				deletedIndex = i;
				break;
			}
		}
		if (deletedIndex >= 0) {
			return true;
		}
		return false;
	}
	static close(messageId) {
		if (this.removeMessage(messageId)) {
			this._NPupBar.refNPupBar.setState({messages: Object.assign({}, this.messages)});
		}
	}
	static reRenderNPupBar() {
		this._NPupBar.refNPupBar.setState({messages: Object.assign({}, this.messages)});
	}
}


src/animations.js
import {NPupDefaults} from './defaults';

export class NPupAnimations {
	static setAnimation() {
		let styleSheet = document.styleSheets;
		let keyframes;
		if (styleSheet[0]) {
			keyframes =
			`@keyframes ${NPupDefaults.lastAnimationName} {
				0% { opacity:1; }
				100% { opacity:0; }
			}`;
			// 99% { opacity:1; display: block; }
			styleSheet[0].insertRule(keyframes, styleSheet[0].cssRules.length);
		}
		if (styleSheet[1]) {
			keyframes =
			`@-webkit-keyframes ${NPupDefaults.lastAnimationName} {
				0% { opacity:1; }
				100% { opacity:0; }
			}`;
			styleSheet[1].insertRule(keyframes, styleSheet[1].cssRules.length);
		}
		if (styleSheet[2]) {
			keyframes =
			`@keyframes ${NPupDefaults.firstAnimationName} {
				0% { opacity:0; max-height: 0px; }
				100% { opacity:1; max-height: 200px; }
			}`;
			styleSheet[2].insertRule(keyframes, styleSheet[2].cssRules.length);
		}
		if (styleSheet[3]) {
			keyframes =
			`@-webkit-keyframes ${NPupDefaults.firstAnimationName} {
				0% { opacity:0; max-height: 0px; }
				100% { opacity:1; max-height: 200px; }
			}`;
			styleSheet[3].insertRule(keyframes, styleSheet[3].cssRules.length);
		}
		// if (styleSheet[1]) {
		// 	keyframes =
		// 	`@-ms-keyframes ${NPupDefaults.lastAnimationName} {
		// 		0% { opacity:1; }
		// 		100% { opacity:0; }
		// 	}`;
		// 	styleSheet[1].insertRule(keyframes, styleSheet[1].cssRules.length);
		// }
		// if (styleSheet[3]) {
		// 	keyframes =
		// 	`@-o-keyframes ${NPupDefaults.lastAnimationName} {
		// 		0% { opacity:1; }
		// 		100% { opacity:0; }
		// 	}`;
		// 	styleSheet[3].insertRule(keyframes, styleSheet[3].cssRules.length);
		// }
		// if (styleSheet[4]) {
		// 	keyframes =
		// 	`@-moz-keyframes ${NPupDefaults.lastAnimationName} {
		// 		0% { opacity:1; }
		// 		100% { opacity:0; }
		// 	}`;
		// 	styleSheet[4].insertRule(keyframes, styleSheet[4].cssRules.length);
		// }
	}
}


src/npupComponent.js
import React, {PropTypes} from 'react';
import {NPupStyles} from './styles';
import {NPupDefaults} from './defaults';

/* React NPupComponent Component */
export class NPupComponent extends React.Component {
	static propTypes = {
		messageId: PropTypes.string,
		text: PropTypes.string,
		closeText: PropTypes.string,
		closeCallback: PropTypes.func,
		timeout: PropTypes.number,
		type: PropTypes.string,
		color: PropTypes.object,
		style: PropTypes.oneOfType([
			PropTypes.object,
			PropTypes.bool
		])
	};
	static isUpdate = false;

	state = {
		contentContainerStyle: null
	};

	getStyles() {
		let styles = {};
		switch (this.props.type) {
			case 'success':
				const successStyle = {
					backgroundColor: NPupStyles.backgroundColor.success,
					color: NPupStyles.textColor.success
				};
				styles.contentContainerStyle = Object.assign({}, NPupStyles.contentContainerStyle, successStyle);
				break;
			case 'error':
				const errorStyle = {
					backgroundColor: NPupStyles.backgroundColor.error,
					color: NPupStyles.textColor.error
				};
				styles.contentContainerStyle = Object.assign({}, NPupStyles.contentContainerStyle, errorStyle);
				break;
			case 'warning':
				const warningStyle = {
					backgroundColor: NPupStyles.backgroundColor.warning,
					color: NPupStyles.textColor.warning
				};
				styles.contentContainerStyle = Object.assign({}, NPupStyles.contentContainerStyle, warningStyle);
				break;
			case 'custom':
				const customStyle = {
					backgroundColor: this.props.color.background
				};
				const customTextStyle = {
					color: this.props.color.text
				};
				styles.contentContainerStyle = Object.assign({}, NPupStyles.contentContainerStyle, customStyle);
				styles.contentTextStyle = Object.assign({}, NPupStyles.contentTextStyle, customTextStyle);
				break;
			default:
				styles.contentContainerStyle = Object.assign({}, NPupStyles.contentContainerStyle);
				break;
		}

		if (!styles.contentTextStyle) {
			styles.contentTextStyle = NPupStyles.contentTextStyle;
		}
		styles.contentButtonContainerStyle = NPupStyles.contentButtonContainerStyle;
		styles.contentButtonStyle = NPupStyles.contentButtonStyle;

		return styles;
	}

	getVisibleState(context) {
		if (this.props.closeText && this.props.timeout <= 0) {
			return;
		}
		let base = this.getStyles().contentContainerStyle;
		const messageId = this.props.messageId;
		setTimeout(function() {
			if (NPupDefaults.isMessageExist(messageId)) {
				context.updateStyle(base, NPupStyles.containerHideStyle);
				setTimeout(function() {
					if (NPupDefaults.isMessageExist(messageId)) {
						NPupDefaults.close(messageId);
					}
				}, NPupDefaults.lastAnimationDuration);
			}
		}, this.props.timeout);
	}

	updateStyle(base, update) {
		this.isUpdate = true;
		this.setState({contentContainerStyle: Object.assign({}, base, update)});
	}

	getBaseStyle() {
		this.isUpdate = true;
		this.setState({contentContainerStyle: this.getStyles().contentContainerStyle});
	}

	componentDidMount() {
		this.getBaseStyle();
		this.getVisibleState(this);
	}

	shouldComponentUpdate() {
		if (this.isUpdate) {
			this.isUpdate = false;
			return true;
		}
		return false;
	}

	closeNPup(context, messageId) {
		if (NPupDefaults.isMessageExist(messageId)) {
			let base = this.getStyles().contentContainerStyle;
			context.updateStyle(base, NPupStyles.containerHideStyle);
			setTimeout(function() {
				if (NPupDefaults.isMessageExist(messageId)) {
					NPupDefaults.close(messageId);
				}
			}, NPupDefaults.lastAnimationDuration);
		}
	}

	actionButton(context, messageId, closeCallback) {
		setTimeout(function() {
			closeCallback();
		}, 0);
		if (NPupDefaults.isMessageExist(messageId)) {
			let base = this.getStyles().contentContainerStyle;
			context.updateStyle(base, NPupStyles.containerHideStyle);
			setTimeout(function() {
				if (NPupDefaults.isMessageExist(messageId)) {
					NPupDefaults.close(messageId);
				}
			}, NPupDefaults.lastAnimationDuration);
		}
	}

	render() {
		let {closeText, closeCallback} = this.props;
		if (closeText) {
			if (closeCallback) {
				return this.renderWithActionButton();
			}
			return this.renderWithButton();
		}
		return this.renderWithoutButton();
	}

	renderWithActionButton() {
		let {messageId, text, closeText, closeCallback, type} = this.props;
		let styles = this.getStyles();
		return (
			<div ref={(r) => this.mainDiv2 = r} className="npup" style={this.state.contentContainerStyle}>
				<div className={type} style={styles.contentTextStyle}>{text}</div>
				<div className={type} style={styles.contentButtonContainerStyle}>
					<div className={type} style={styles.contentButtonStyle} onClick={this.actionButton.bind(this, this, messageId, closeCallback)}>{closeText}</div>
				</div>
			</div>
		);
	}

	renderWithButton() {
		let {messageId, text, closeText, type} = this.props;
		let styles = this.getStyles();
		return (
			<div ref={(r) => this.mainDiv2 = r} className="npup" style={this.state.contentContainerStyle}>
				<div className={type} style={styles.contentTextStyle}>{text}</div>
				<div className={type} style={styles.contentButtonContainerStyle}>
					<div className={type} style={styles.contentButtonStyle} onClick={this.closeNPup.bind(this, this, messageId)}>{closeText}</div>
				</div>
			</div>
		);
	}

	renderWithoutButton() {
		let {text, type} = this.props;
		let styles = this.getStyles();
		return (
			<div ref={(r) => this.mainDiv2 = r} className="npup" style={this.state.contentContainerStyle}>
				<div className={type} style={styles.contentTextStyle}>{text}</div>
			</div>
		);
	}
}


src/styles.js
import {NPupDefaults} from './defaults';

export class NPupStyles {
	static backgroundColor = {
		default: '#000000',
		warning: '#000000',
		error: '#000000',
		success: '#000000'
	};
	static textColor = {
		default: '#FFFFFF',
		warning: '#FFFFFF',
		error: '#FFFFFF',
		success: '#FFFFFF'
	};
	static buttonColor = {
		default: '#FFCC00',
		warning: '#FFCC00',
		error: '#FFCC00',
		success: '#FFCC00'
	};
	static containerStyle = {
		position: 'fixed',
		width: '50%',
		margin: '0px 0px 24px 24px',
		right: '0px',
		bottom: '0px',
		left: '0px',
		textAlign: 'center',
		zIndex: '999999',
		pointerEvents: 'none'
	};
	static contentContainerStyle = {
		cursor: 'default',
		display: 'table',
		minWidth: '320px',
		maxWidth: '512px',
		maxHeight: '200px',
		borderRadius: '3px',
		backgroundColor: NPupStyles.backgroundColor.default,
		color: NPupStyles.textColor.default,
		padding: '2px 16px 2px 24px',
		margin: '5px 0px 0px 0px',
		pointerEvents: 'all',
		transition: 'all ' + NPupDefaults.firstAnimationDuration + 'ms ease',
		// Vendor Prefixes
		msTransition: 'all ' + NPupDefaults.firstAnimationDuration + 'ms ease',
		WebkitTransition: 'all ' + NPupDefaults.firstAnimationDuration + 'ms ease',
		OTransition: 'all ' + NPupDefaults.firstAnimationDuration + 'ms ease',
		MozTransition: 'all ' + NPupDefaults.firstAnimationDuration + 'ms ease',
		animationName: NPupDefaults.firstAnimationName,
		animationTimingFunction: 'ease',
		animationDuration: NPupDefaults.firstAnimationDuration + 'ms',
		animationDelay: '0.0s',
		animationIterationCount: 1,
		animationDirection: 'normal',
		animationFillMode: 'forwards'
	};
	static contentTextStyle = {
		cursor: 'text',
		float: 'left',
		padding: '14px 0px 14px 0px',
		margin: '0px',
		textAlign: '-webkit-left',
		fontSize: '13px',
		color: NPupStyles.textColor.default
	};
	static contentButtonContainerStyle = {
		cursor: 'pointer',
		float: 'right',
		display: 'table',
		height: '36px',
		margin: '3.5px 0px',
		padding: '0px 6px'
	};
	static contentButtonStyle = {
		display: 'table-cell',
		verticalAlign: 'middle',
		fontSize: '14px',
		fontWeight: '500',
		color: NPupStyles.buttonColor.default
	};
	static containerHideStyle = {
		animationName: NPupDefaults.lastAnimationName,
		animationTimingFunction: 'ease-in-out',
		animationDuration: NPupDefaults.lastAnimationDuration + 'ms',
		animationDelay: '0.0s',
		animationIterationCount: 1,
		animationDirection: 'normal',
		animationFillMode: 'forwards'
	};

	static changeBackgroundColorStyle(newStyle) {
		this.backgroundColor = Object.assign({}, this.backgroundColor, newStyle);
	}

	static changeTextColorStyle(newStyle) {
		this.textColor = Object.assign({}, this.textColor, newStyle);
	}

	static changeButtonColorStyle(newStyle) {
		this.buttonColor = Object.assign({}, this.buttonColor, newStyle);
	}

	static changeContainerStyle(newStyle) {
		this.containerStyle = Object.assign({}, this.containerStyle, newStyle);
	}

	static changeContentContainerStyle(newStyle) {
		this.contentContainerStyle = Object.assign({}, this.contentContainerStyle, newStyle);
	}

	static changeContentTextStyle(newStyle) {
		this.contentTextStyle = Object.assign({}, this.contentTextStyle, newStyle);
	}

	static changeContentButtonContainerStyle(newStyle) {
		this.contentButtonContainerStyle = Object.assign({}, this.contentButtonContainerStyle, newStyle);
	}

	static changeContentButtonStyle(newStyle) {
		this.contentButtonStyle = Object.assign({}, this.contentButtonStyle, newStyle);
	}

	static changeContainerHideStyle(newStyle) {
		this.containerHideStyle = Object.assign({}, this.containerHideStyle, newStyle);
	}
}
